perceval/utils/density_matrix.py

```diff
@@ -355,14 +356,104 @@ def normalize(self):
         factor = self.mat.trace()
         self.mat = (1/factor)*self.mat
 
-    def sample(self, count: int = 1):
+    def sample(self, count: int = 1) -> BSSamples:
         """
         sample on the density matrix
         """
         self.normalize()
-        output = random.choices(self.reverse_index, list(self.mat.diagonal()), k=count)
+        samples = random.choices(self.reverse_index, list(self.mat.diagonal()), k=count)
+        output = BSSamples()
+        for state in samples:
+            output.append(state)
         return output
 
+    def measure(self, modes: Union[list[int], int], all_results: bool = True):
+        """
+        makes a measure on a list of modes
+        :param modes: a list of integer for the modes you want to measure
+        :param all_results: whether you want a resulting mixed state or a simple sample
+        """
+        self.normalize()
+        if isinstance(modes, int):
+            modes = [modes]
+
+        if all_results:
+            projectors = self._construct_all_projectors(modes)
+            res = dict() # result fo the form {measured FockState: (remaining density matrix, probability)
+            for key_fs, item_list in projectors.items():
+                basis = item_list[0] # FockBasis of possible measurement
+                projector = item_list[1]
+                prob = item_list[2]
+                if prob != 0:
+                    collapsed_dm = projector @ self.mat @ projector.T  # wave function collapse
+                    resulting_dm = DensityMatrix(collapsed_dm, basis)
+                    resulting_dm.normalize()
+                    res[key_fs] = (prob, resulting_dm)
+            return res
+        else:
+            sample = self.sample()[0]  # if you want to sample instead of keeping all the possibilities
+            measure, remaining = self._divide_fock_state(sample, modes)
+            basis, proj = self._construct_projector(modes, measure)
+            return DensityMatrix(proj @ self.mat @ proj.T, basis)
+
+    def _construct_projector(self, modes, fock_state) -> tuple[FockBasis, dok_array]:
+        """
+        Construct the projection operator onto the subspace of some number photons on some mode
+        """
+        if len(modes) != fock_state.m:
+            raise ValueError(f"you can't have {fock_state} in  {len(modes)} number of modes")
+
+        basis = FockBasis(self.m-fock_state.m, self.n_max-fock_state.n)
+        projector = dok_array((len(basis), self.size), dtype=float)
+
+        for i, fs in enumerate(self.reverse_index):
```

Comment:
I agree that the name is confusing. Maybe "inverse" would be better because if you see it as a mapping between fockstates and indexes, one would be the inverse of the other. In any case this is something I defined a while back and if I change it I would rather do it in a separated ticket to avoid conflict with other branches.